### PR TITLE
WS connection fix

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

Currently, our WebSocket connection breaks on core restart and about 1 after startup, this is silent since we never logged anything.

This PR implements a listener and more error handling so when the connection is dropped we can reinstate the client.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

File: supervisor/homeassistant/websocket.py [hunk 1]
```diff
@@ -22,61 +26,112 @@ class WSClient:
...
 
     @classmethod
     async def connect_with_auth(
-        cls, session: aiohttp.ClientSession, url: str, token: str
+        cls, session: aiohttp.ClientSession, loop, url: str, token: str
     ) -> "WSClient":
         """Create an authenticated websocket client."""
         try:
-            client = await session.ws_connect(url, ssl=False)
+            client = await session.ws_connect(url, timeout=500)
```
Comment (pvizeli) [inline supervisor/homeassistant/websocket.py:121]: We need disable SSL check for connection to core

File: supervisor/homeassistant/websocket.py [hunk 2]
```diff
@@ -22,61 +26,112 @@ class WSClient:
...
+        client: aiohttp.ClientWebSocketResponse,
     ):
         """Initialise the WS client."""
-        self.ha_version: AwesomeVersion = ha_version
-        self.client: aiohttp.ClientWebSocketResponse = client
-        self.message_id: int = 0
-        self._lock: asyncio.Lock = asyncio.Lock()
+        self.ha_version = ha_version
+        self._client = client
+        self._message_id: int = 0
```
Comment (pvizeli) [inline supervisor/homeassistant/websocket.py:37]: I think that gets messed up on the restart of the supervisor during the runtime? Maybe would be better to use https://github.com/home-assistant/core/blob/dev/homeassistant/util/uuid.py#L12

Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:37]: I have restarted core and SU multiple times with this, there were no issues detected so far

File: supervisor/homeassistant/websocket.py (hunk 1 above)
Comment (agners) [inline supervisor/homeassistant/websocket.py]: So this sets `Timeout for websocket to close` to 8.3 minutes, any specific reason for that number? Do we have a guarantee to have a message within this time? Should we set it to `None`?

File: supervisor/homeassistant/websocket.py [hunk 3]
```diff
@@ -22,61 +26,112 @@ class WSClient:
...
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
+
+        if msg.type != WSMsgType.TEXT:
+            raise HomeAssistantWSError(f"Received non-Text message: {msg.type}")
```
Comment (agners) [inline supervisor/homeassistant/websocket.py:98]: This seems an error which which is related to the message itself, I think we should pass it to the caller via `future.set_exception(...)`.

File: supervisor/homeassistant/websocket.py [hunk 4]
```diff
@@ -22,61 +26,112 @@ class WSClient:
...
+
+        try:
+            data = msg.json()
+        except ValueError as err:
+            raise HomeAssistantWSError(f"Received invalid JSON - {msg}") from err
```
Comment (agners) [inline supervisor/homeassistant/websocket.py:103]: same here, I think this exception should be passed to the future.

File: supervisor/homeassistant/websocket.py (hunk 1 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py]: This is outdated

File: supervisor/homeassistant/websocket.py (hunk 3 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:98]: At this point we don't know if we have something waiting, and we can't know until we read the message data further down

File: supervisor/homeassistant/websocket.py (hunk 4 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:103]: At this point we don't know if we have something waiting, and we can't know until we read the message data further down

File: supervisor/homeassistant/websocket.py (hunk 2 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:37]: Home Assistant also expect an int for "id" and does not allow for any additional fields that we could use for uuid

File: supervisor/homeassistant/websocket.py (hunk 4 above)
Comment (agners) [reply supervisor/homeassistant/websocket.py:103]: Oh I see, response ID is part of the payload, which we need to get the future. Yeah well then.

File: supervisor/homeassistant/websocket.py [hunk 5]
```diff
@@ -22,41 +26,95 @@ class WSClient:
...
+        if data["type"] == "result":
+            if (future := self._futures.get(data["id"])) is None:
+                return
+
+            if data["success"]:
+                future.set_result(data["result"])
+                return
 
-            raise HomeAssistantWSError(response)
+            _LOGGER.error("Unsuccessful websocket message - %s", data)
```
Comment (agners) [inline supervisor/homeassistant/websocket.py:113]: But I think in this case we could inform the future about the unsuccessful message. I guess setting an exception would be most fitting here.

Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:113]: Yeah, we could do that :+1:

File: supervisor/homeassistant/websocket.py [hunk 6]
```diff
@@ -22,41 +26,97 @@ class WSClient:
...
+        msg = await self._client.receive()
+        _LOGGER.debug("Received: %s", msg)
+
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
```
Comment (agners) [inline supervisor/homeassistant/websocket.py:95]: It seems to me we throw graceful connection closes and errors into the same bucket here. Can we differentiate that?

I guess `CLOSE` is the message we get if we are dealing with a regular shutdown.

How about:
```
if msg.type == WSMsgType.CLOSE:
    return False

if msg.type in (
    WSMsgType.CLOSED,
    WSMsgType.CLOSING,
    WSMsgType.ERROR,
):
    raise HomeAssistantWSConnectionError()
```

File: supervisor/homeassistant/websocket.py [hunk 7]
```diff
@@ -22,41 +26,97 @@ class WSClient:
...
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
```
Comment (agners) [inline supervisor/homeassistant/websocket.py:76]: .. then we can do something better here:

```
            while self.connected:
                if not await self._receive_json():
                    _LOGGER.debug("Websocket connection closed gracefully.");
                    break
        except HomeAssistantWSConnectionError:
             _LOGGER.warning("Websocket connection closed with connection error.");
```

File: supervisor/homeassistant/websocket.py (hunk 6 above)
Comment (agners) [reply supervisor/homeassistant/websocket.py:95]: (needs also a `return True` at the end of the function)

Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:95]: We can, but that adds logic that is not needed.
In any of these scenarios, we need to start the process of closing down and recreating the client, which should be transparent for the user.

Comment (agners) [reply supervisor/homeassistant/websocket.py:95]: Yes, we need to recreate the client in any case.

But I'd still prefer if we differentiate between expected flow (regular close) and error case. I am mainly interested in having a warning in the log or something in the error case. This helps us to understand if the system is behaving as we expect.

File: supervisor/homeassistant/websocket.py (hunk 7 above)
Comment (agners) [reply supervisor/homeassistant/websocket.py:76]: It is mainly that `pass` which I don't like. We could also just add a debug message here, but as stated above, I think its worth differentiate these cases. After all this is going to be a ever more important communication channel, having it sold is worth the effort IMHO :)

File: supervisor/homeassistant/websocket.py (hunk 6 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:95]: I don't agree, but whatever. I'll add something

**Review by agners — APPROVED** [review]

**Pull request merged**